dask/array/core.py

```diff
@@ -2833,20 +2833,21 @@ def where(condition, x=None, y=None):
     if x is None or y is None:
         raise TypeError(where_error_message)
 
-    x = asarray(x)
-    y = asarray(y)
-
-    shape = broadcast_shapes(x.shape, y.shape)
-    dtype = np.promote_types(x.dtype, y.dtype)
+    if np.isscalar(condition):
+        # Match the dtype inference logic from elemwise
+        vals = [np.empty((1,) * a.ndim, dtype=a.dtype)
+                if not is_scalar_for_elemwise(a) else a
+                for a in (x, y)]
+        dtype = apply_infer_dtype(np.where, [condition] + vals, {}, 'where', suggest_dtype=False)
```

Comment:
> Using `promote_types` gives the wrong type if one of the inputs (x or y) is a scalar, because the numpy (and `elemwise`) promotion rules are different for scalars.

Could you please explain why this is relevant to this case? We are calling `promote_types` on the `dtype`s themselves not on the arrays or scalars. So dimensionalities of the objects should be irrelevant.

> As for unconditional `broadcast_shapes`, try running the following on 0.15.1 and looking at the mydask.png that it produces...

There should be no graph coming from `broadcast_shapes`. This merely returns the expected `shape`. If you are referring to `broadcast_to` or `astype`, these again should be no-ops if everything shares the intended `shape` and `dtype` already. So if you are getting new things in the graph from your test code, I for one would be very curious to see the graph. Please share.